packages/proto-loader/bin/proto-loader-gen-types.ts

```diff
@@ -0,0 +1,745 @@
+          formatter.writeLine(`${name}(metadata: grpc.Metadata, options: grpc.CallOptions, callback: ${callbackType}): ${callType};`);
+          formatter.writeLine(`${name}(metadata: grpc.Metadata, callback: ${callbackType}): ${callType};`);
+          formatter.writeLine(`${name}(options: grpc.CallOptions, callback: ${callbackType}): ${callType};`);
+          formatter.writeLine(`${name}(callback: ${callbackType}): ${callType};`);
+        }
+      } else {
+        if (method.responseStream) {
+          // Server streaming
+          const callType = `grpc.ClientReadableStream<${responseType}>`;
+          formatter.writeLine(`${name}(argument: ${requestType}, metadata: grpc.Metadata, options?: grpc.CallOptions): ${callType};`);
+          formatter.writeLine(`${name}(argument: ${requestType}, options?: grpc.CallOptions): ${callType};`);
+        } else {
+          // Unary
+          const callType = 'grpc.ClientUnaryCall';
+          formatter.writeLine(`${name}(argument: ${requestType}, metadata: grpc.Metadata, options: grpc.CallOptions, callback: ${callbackType}): ${callType};`);
+          formatter.writeLine(`${name}(argument: ${requestType}, metadata: grpc.Metadata, callback: ${callbackType}): ${callType};`);
+          formatter.writeLine(`${name}(argument: ${requestType}, options: grpc.CallOptions, callback: ${callbackType}): ${callType};`);
+          formatter.writeLine(`${name}(argument: ${requestType}, callback: ${callbackType}): ${callType};`);
+        }
+      }
+    }
+    formatter.writeLine('');
+  }
+  formatter.unindent();
+  formatter.writeLine('}');
+}
+
+function generateServiceHandlerInterface(formatter: TextFormatter, serviceType: Protobuf.Service, options: GeneratorOptions) {
+  if (options.includeComments) {
+    formatComment(formatter, serviceType.comment);
+  }
+  formatter.writeLine(`export interface ${serviceType.name}Handlers extends grpc.UntypedServiceImplementation {`);
+  formatter.indent();
+  for (const methodName of Object.keys(serviceType.methods).sort()) {
+    const method = serviceType.methods[methodName];
+    if (options.includeComments) {
+      formatComment(formatter, method.comment);
+    }
+    const requestType = getTypeInterfaceName(method.resolvedRequestType!) + '__Output';
+    const responseType = getTypeInterfaceName(method.resolvedResponseType!);
+    if (method.requestStream) {
+      if (method.responseStream) {
+        // Bidi streaming
+        formatter.writeLine(`${methodName}(call: grpc.ServerDuplexStream<${requestType}, ${responseType}>): void;`);
+      } else {
+        // Client streaming
+        formatter.writeLine(`${methodName}(call: grpc.ServerReadableStream<${requestType}, ${responseType}>, callback: grpc.sendUnaryData<${responseType}>): void;`);
+      }
+    } else {
+      if (method.responseStream) {
+        // Server streaming
+        formatter.writeLine(`${methodName}(call: grpc.ServerWritableStream<${requestType}, ${responseType}>): void;`);

```

Comment:
Due to a [mismatch of types](https://github.com/grpc/grpc-node/issues/1581), would it better to define the handlers using these types:

- grpc.handleUnaryCall
- grpc.handleServerStreamingCall
- grpc.handleClientStreamingCall
- grpc.handleBidiStreamingCall

This would provide better compatibility between between the `grpc` and `@grpc/grpc-js` packages, allowing you to use this TypeScript generator with the legacy `grpc` package. (See also https://github.com/grpc/grpc-node/pull/1587)

Reply:
See https://github.com/murgatroid99/grpc-node/pull/2